_posts/2025-02-28-a-practical-guide-for-selecting-HNSW-hyperparameters.md

```diff
@@ -0,0 +1,683 @@
+| [Last.fm](http://millionsongdataset.com/lastfm/)        | 65           | 292,385      | 50,000      |         100 | Inner product | Matrix factorization                           | Song recommendation           |                                                   
+| [Word2bits](https://github.com/agnusmaximus/Word2BitsD199572657/)      | 800          | 399,000      | 1,000       |         100 | Hamming       | Word2Vec with quantized parameter              | Language, English Wikipedia   |                                                       
+| [GIST](http://corpus-texmex.irisa.fr/)         | 960          | 1,000,000    | 1,000       |         100 | Euclidean     | GIST descriptors, INRIA C implementation       | Image                         |                                                        
+| [MS MARCO](https://microsoft.github.io/msmarco/)        | 384          | 1,000,000    | 50,000      |         100 | Euclidean     | MiniLLM                                        | Language, Question answering  |                                                           
+| [openai-dbpedia](https://huggingface.co/datasets/Qdrant/dbpedia-entities-openai3-text-embedding-3-large-1536-1M) | 1,536        | 950,000      | 50,000      |         100 | Euclidean     | text-embedding-3-large                         | Language, DBPedia             | 
+
+
+For each dataset, we evaluated a grid of 80 HNSW configurations based on the following search space:
+
+```python
+search_space = {
+    "M": [8, 16, 32, 64, 128],
+    "efConstruction": [32, 64, 128, 256],
+    "efSearch": [32, 64, 128, 256]
+}
+```
+
+For these experiments, we used an OpenSearch 2.15 cluster with three cluster manager nodes and six data nodes, each running on an `r6g.4xlarge.search` instance. We evaluated test vectors in batches of 100 and recorded query throughput and recall@10 for each HNSW configuration. In the next section, we'll introduce the algorithm used to learn the portfolio.  
+
+### Method  
+
+To capture different trade-offs between recall and throughput, we used a simple linearization approach, assigning values between 0 and 1, inclusive, to both recall and throughput. Given a specific weighting, we identified the configuration that maximizes the linearized object using the following four steps:  
+
+1. **Normalize recall and throughput** – Apply min-max scaling within each dataset so that recall and throughput values are comparable.  
+2. **Compute weighted metric** – Combine the normalized recall and throughput using the assigned weights into a new weighted metric.  
+3. **Average across datasets** – Calculate the average weighted metric across datasets.  
+4. **Select the best configuration** – Identify the configuration that maximizes the average weighted metric.  
+
+The following figure illustrates the algorithm using an example with two datasets and three configurations.
```

Comment:
"The following figure illustrates the use of the algorithm with an example containing two datasets and three configurations"?

Reply:
I think the original is clearer.